### PR TITLE
fix(compiler): apply style on :host attribudes in prod builds.

### LINKED ISSUE #49100: CSS styles not applied if using attribute HostBinding and styling a DIV with the direct child selector in a production build
### Which @angular/* package(s) are the source of the bug?

Don't known / other

### Is this a regression?

Yes

### Description

If using attribute HostBinding and styling a DIV element with the direct child selector, the styles are not applied.
**This happens only in a producton build.**

Reproduction:
**app.component.html**
```html
<div class="example-1">
  Example 1 (css: > div)
</div>
<div class="example-2">
  Example 2 (css: div)
</div>
<section class="example-3">
  Example 3 (css: > section)
</section>
<section class="example-4">
  Example 4 (css: section)
</section>
```
**app.component.ts**
```ts
import {Component, HostBinding} from '@angular/core';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent {

  @HostBinding('attr.data-foo')
  public foo = 'bar';
}

```
**app.component.scss**
```scss
:host {
  &[data-foo="bar"] {
    > div.example-1 {
      background-color: cyan; // not working
    }

    div.example-2 {
      background-color: yellow; // working
    }

    > section.example-3 {
      background-color: red; // working
    }

    section.example-4 {
      background-color: green; // working
    }
  }
}

```
image

In the transpiled CSS we noticed various issues:

Example 1:

- "nghost" attribute is missing after the attribute selector _[data-foo=bar]_
- "nghost" attribute after _div.example-1_ should instead be "ngcontent"

Example 2:

- "ngcontent" attribute is missing after _div.example-2_. **IMPORTANT! This is not the cause of this problem, but causes the styles to leak into child components. Please let me know if I should file a separate issue for this case.**

transpiled css

### Please provide a link to a minimal reproduction of the bug

https://github.com/Marcarrian/angular-15-attribute-host-binding-issue

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 15.1.5        
Node: 16.16.0              
Package Manager: npm 8.11.0
OS: win32 x64              
                           
Angular: 15.1.4            
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1501.5
@angular-devkit/build-angular   15.1.5
@angular-devkit/core            15.1.5
@angular-devkit/schematics      15.1.5
@angular/cli                    15.1.5
@schematics/angular             15.1.5
rxjs                            7.8.0
typescript                      4.9.5
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JeanMeche**
In prod builds, selectors are optimized and spaces a removed. #48558 introduced a regression on selectors without spaces. This commit fixes this.

Fixes #49100

Currently this PR blocks #49169. 

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

**Comment by JeanMeche** [comment]
I wouldn't expect this commit to trigger regressions. Are the failling internal tests a false positive ?

**Comment by JeanMeche** [comment]
@dylhunn How did the presubmit go ?

**Comment by alan-agius4** [comment]
[Presubmit](http://fusion2/presubmit/539094397/OCL:539094397:BASE:539096525:1686326599821:5f8fe0cd/targets)

**Comment by dylhunn** [comment]
Just checked Alan's new presubmit. It's clean, modulo 12 flakes and preexisting failures. So we should be good to go in that regard.

File: packages/compiler/src/shadow_css.ts [hunk 1]
```diff
@@ -713,7 +713,7 @@ export class ShadowCss {
       // (ie: ".\fc ber" for ".über") is not a separator between 2 selectors
       // also keep in mind that backslashes are replaced by a placeholder by SafeSelector
       // These escaped selectors happen for example when esbuild runs with optimization.minify.
-      if (part.match(_placeholderRe) && selector[res.index + 1]?.match(/[a-fA-F\d]/)) {
+      if (part.match(/__esc-ph-(\d+)__/g) && selector[res.index + 1]?.match(/[a-fA-F\d]/)) {
```
Comment (JoostK) [inline packages/compiler/src/shadow_css.ts:716]: I don't think the capturing group is needed here, nor is the global flag:

```suggestion
      if (part.match(/__esc-ph-\d+__/) && selector[res.index + 1]?.match(/[a-fA-F\d]/)) {
```

Without the global flag this regex can be safely hoisted to stored in a top-level variable.

File: packages/compiler/src/shadow_css.ts [hunk 2]
```diff
@@ -821,12 +827,10 @@ const _colonHostContextRe = /:host-context/gim;
 
 const _newLinesRe = /\r?\n/g;
 const _commentRe = /\/\*[\s\S]*?\*\//g;
-const _commentWithHashRe = /\/\*\s*#\s*source(Mapping)?URL=/g;
+const _commentWithHashRe = /\/\*\s*#\s*source(Mapping)?URL=[\s\S]+?\*\//g;
```
Comment (JoostK) [inline packages/compiler/src/shadow_css.ts:830]: Is this change related to this fix? It doesn't seem to be the case and there are no test updates, so I guess this should be reverted?

File: packages/compiler/src/shadow_css.ts (hunk 1 above)
Comment (JoostK) [reply packages/compiler/src/shadow_css.ts:716]: Does this work when `part` contains both an escaped pseudo class _and_ an escaped backslash? The regex test is across all of `part`, so wouldn't the escaped backslash escaping then still cause this logic to be triggered for the pseudo occurrence, resulting in the same issue as before? I may be missing something here, but it'd at least be interesting to have a test for such case.

Comment (JeanMeche) [reply packages/compiler/src/shadow_css.ts:716]: Yep I can confirm that a selector like `:host.pr\\fc fung` isn't handled correctly.

Comment (JeanMeche) [reply packages/compiler/src/shadow_css.ts:716]: But `.one\\:\\fc ber {}` is handled correctly I'm adding it in the tests.
 
I won't fix the :host with an escaped selector, it will be hard to fix.

File: packages/compiler/test/shadow_css/host_and_host_context_spec.ts [hunk 3]
```diff
@@ -27,6 +27,17 @@ describe('ShadowCss, :host and :host-context', () => {
...
 
+    it('should handle attribute and next operator without spaces', () => {
+      expect(shim(':host[foo]>div {}', 'contenta', 'a-host'))
+          .toEqualCss(('[foo][a-host] > div[contenta] {}'));
+    });
+
+    // we know that the following test doesn't pass
+    // the host attribute is added before the space
+    xit('should handle host with escaped class selector', () => {
+      expect(shim(':host.pr\\fc fung {}', 'contenta', 'a-host')).toEqual('.pr\\fc fung[a-host] {}');
```
Comment (JoostK) [inline packages/compiler/test/shadow_css/host_and_host_context_spec.ts:38]: Could this input be accompanied by a comment that specifies what the intent is here, i.e. the escape sequence in `\\fc fung` actually represents `.prüng` (if I decoded that correctly, but that's exactly why a comment here would help)

**Review by JoostK — APPROVED** [review]
Overall I think it's time to take a step back and maybe revisit some of the approach here, but this change is in the right direction. Left two minor comments but otherwise LGTM.

**Comment by JoostK** [comment]
Labelling as risky since this might affect rendering results in a subtle way. I recommend syncing this separately once TGP comes back green.

**Comment by JoostK** [comment]
@JeanMeche  could you also fix the typo in the commit message (attributes is misspelled)

**Comment by dylhunn** [comment]
One more presubmit in cl/572379101

**Comment by dylhunn** [comment]
The global presubmit shows one broken cloud console test. I have reached out to the code owners to try to get their help fixing it.

**Comment by dylhunn** [comment]
The owning team confirms the g3 failure as a flake.

Caretaker: TGP is green, good to merge.

**Comment by atscott** [comment]
This PR was merged into the repository by commit 0198d21231565f5eeaa27a871b9bb9e950b4a869.
